### ISSUE TITLE
POST method XMLHttpRequest request header Origin is null

### ISSUE THREAD

**Issue by jesus2099** [greasemonkey/greasemonkey#3111]
Greasemonkey 4.11 in Firefox 86.0

I don’t know why but XMLHttpRequest method POST requests made by userscripts in Greasemonkey, are sent with Origin header _set_ to `null`, even when I am requesting the same host.

Here is **how to reproduce**: https://github.com/greasemonkey/greasemonkey/issues/3111#issuecomment-797091459

Shouldn’t Origin header be equal to `location.protocol + "//" + location.host` to let server manage CORS policy?

**Comment by jesus2099** [comment]
I created a simple test userscript: https://gist.github.com/jesus2099/b4d8584d761c42c43da8d0cf3e5d2a62

[Install it](https://gist.github.com/jesus2099/b4d8584d761c42c43da8d0cf3e5d2a62/raw/cc713135bd0436b7bb97ff7077fe53775acf4d17/origin-test.user.js) and go to https://httpbin.org (you can try `http` and `https` with same result).
See the alert content, after 1 second: the Origin request header is set to `null` although we are doing a post request on the same origin (httpbin.prg/post).

## Greasemonkey

    Origin: null

## Violentmonkey and Tampermonkey

    Origin: https://httpbin.org
    Referer: http://httpbin.org/

**Comment by jesus2099** [comment]
Indeed it works now for me too with Greasemonkey 4.11 in Firefox 87.0.

I don't know if Origin is now not empty as there is https://github.com/greasemonkey/greasemonkey/issues/3110 preventing me debugging.
Origin is still empty on my small test https://github.com/greasemonkey/greasemonkey/issues/3111 but maybe MBS did some cool changes.

**Comment by jesus2099** [comment]
I don't remember why I closed this.
The issue is still there, no Origin set with Firefox and Greasemonkey with that test above.

**Comment by afontenot** [comment]
> Shouldn’t Origin header be equal to location.protocol + "//" + location.host to let server manage CORS policy?

No. As per Mozilla's [documentation for content scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts):

> Content scripts get the same cross-domain privileges as the rest of the extension: so if the extension has requested cross-domain access for a domain using the permissions key in manifest.json, then its content scripts get access that domain as well.
>
> This is accomplished by exposing more privileged XHR and fetch instances in the content script, which has the side-effect of not setting the Origin and Referer headers like a request from the page itself would; this is often preferable to prevent the request from revealing its cross-origin nature.

Unlike in Chrome, content scripts in Firefox are not subject to CORS. There's a workaround mentioned in the documentation if you do need requests sent in the page context.

**Comment by jesus2099** [comment]
I was talking about userscripts, I don't know what are content scripts, exactly.
For me, userscripts have always run as if from the page/website itself, not _external_ like this doc says for content scripts.

It's not a Firefox specific issue that I noticed and reported, as the other two userscript engines work the same way (Origin not null), cross-browser.

For the record, the work-around in this doc is using `window.eval()`.

**Comment by afontenot** [comment]
Userscript managers can inject scripts in two basic ways: either directly into the page, or as a content script, which is a feature for running Javascript in a page for browser extensions that's part of WebExtensions. The latter (content scripts) is sandboxed by the browser and considered the more modern and secure way to do this. Greasemonkey uses it exclusively. 

Violentmonkey injects your script in the page by default, but this breaks on pages with strong [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) that disallows inline scripts. Firefox respects the page CSP and blocks such scripts even when they are added by an extension. Script writers can opt in to using content scripts on Violentmonkey by setting `// @inject-into content`. For scripts that don't set their context manually in this way, the user's choice of default will decide, but the default setting for this is the page context.

> For the record, the work-around in this doc is using window.eval().

You're mistaken. The page says that certain operations can be performed in the page context by content scripts by using `content`, e.g.

    var xhr = new XMLHttpRequest();
    if (content && content.XMLHttpRequest) {
      xhr = new content.XMLHttpRequest();
    }

Using this in your example code works for me, I see `"Origin": "https://httpbin.org"` in the alert popup.

**Comment by jesus2099** [comment]
Wow.
Thank you very much, I save this email for later reference.
Userscript is becoming less and less trivial, with all these layers and also with all websites that use many JavaScript libraries.
I will surely need your great info someday.
So there is no bug, here, right?
I let the developers close this bug because I am not sure myself.

**Comment by afontenot** [comment]
No worries, I'm not a dev by the way, just happened to see your issue when I was filing my own and knew the answer. You're right that user scripts have become more complicated, the various managers have all taken different approaches for handling script injection, a result of the browsers tightening up security. 

I don't think there's a bug here. The "real" issue is just the difficulty of writing code that works on all browsers and script managers, but IDK if Greasemonkey can do anything about that while sticking to content scripts as its injection method.

**Comment by Sxderp** [comment]
I'm going to mention that this "issue" will go away when/if [1578405](https://bugzilla.mozilla.org/show_bug.cgi?id=1578405) is implemented. When implemented all XHR / fetch requests will be made with the page context and will have the page Origin / Referrer / whatever. In order to perform cross site you will need to send the request to the background via GM.XHR.

--

This is getting a bit off topic and the discussion should probably end here, but I wanted to chime in on one more item. There's another way and that's to use the "user script API". This is exclusive to FireFox and is like a sandboxed version of the "content script" that @afontenot mentioned. I believe [FireMonkey](https://addons.mozilla.org/en-US/firefox/addon/firemonkey/) uses this exclusively.

**Comment by afontenot** [comment]
@sxderp is there any hope of Greasemonkey ever supporting the userScript API? There has been a bug report about it since 2019: https://github.com/greasemonkey/greasemonkey/issues/3028 

Maybe I'm misreading the bug report you link, but it sounds like they're deprecating cross-origin requests entirely? Is it just confusingly worded and they mean that CORS will be applied to content scripts in the future?

AFAIK some of the Greasemonkey devs were involved in its creation but now seem rather skeptical that it can do everything you / they need? Someone might need to do a "state of the project" type bug report for Greasemonkey, anyway. Development seems sort of dead.

**Comment by Sxderp** [comment]
CORS will be applied. Requests will be treated as if done by the page. They will fail unless the destination server responds with the appropriate CORS headers.

This can be worked around by performing the request using the background page. Greasemonkey offers this functionality through the GM.XHR interface. I do not know about other user-script tools.

**Comment by jesus2099** [comment]
For readability and compatibility reasons, I usually use JavaScript with no libraries, like `var xhr = new XMLHttpRequest();` in this case.

Anyway for my script, [the issue went away by magic](https://github.com/jesus2099/konami-command/issues/546#issuecomment-814223297), maybe thanks to a server change.

**Comment by arantius** [comment]
The fact that the extension's content script cross-site XHR has leaked to you is an accidental/undocumented feature.  If you want cross origin requests to work, and you want support making/keeping them working, you should use https://wiki.greasespot.net/GM.xmlHttpRequest .

**Comment by jesus2099** [comment]
I don't know.
For me, started with Opera's User JS, it's normal that user script runs as if in the page.
And I prefer not using libraries for basic things, for compatibility purpose.

**Comment by jesus2099** [comment]
Thank you very much for your feedback.

I realise now that I recently changed my script that had always been a `//@grant none` script before [this commit](https://github.com/jesus2099/konami-command/commit/c60ee8134126311bae881a155657964785102612#diff-77fdf393707a3ae2a5416e8a1cc01c14839892c3334b9746712cd3cd46eb8421).

Maybe this strange behaviour happens because it’s no longer a `//@grant none` script. Those have special good treatments, if I remember correctly.

I’ll revert my `//@grant GM_info` and see if it’s better.

**Comment by Cerberus-tm** [comment]
@jesus2099 Don't you mean `GM.Info`?

**Comment by jesus2099** [comment]
I do use `GM_info`, not `GM.info`, indeed.

I only use GM underscore functions because I don’t understand `promise` and all that new stuff, and probably never will.

But I’ve only been testing my userscripts with Violentmonkey, for a long time, now.

**Comment by Cerberus-tm** [comment]
Unfortunately, I believe Greasemonkey disabled all the `GM_` functions and now only uses the `GM.` stuff with promises. I agree that it is hard to work with, so much harder than the old method. And changing existing scripts to use the new method is a nightmare. As a result, I needed to switch to Violentmonkey too, for many scripts. Too many very important scripts were broken in Greasemonkey. But we cannot change this decision, unfortunately.
